src/Symfony/Component/Security/Core/Exception/CustomAuthenticationException.php

```diff
@@ -0,0 +1,92 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Core\Exception;
+
+/**
+ * An authentication exception where you can add a message that will
+ * be shown to the user, without needing to create a sub-class.
+ *
+ *  throw CustomAuthenticationException::createWithSafeMessage(
+ *      'That was a ridiculous username'
+ *  );
+ *
+ * @author Ryan Weaver <ryan@knpuniversity.com>
+ */
+class CustomAuthenticationException extends AuthenticationException
+{
+    private $messageKey;
+
+    private $messageData;
+
+    /**
+     * Helper method to create this exception and set the safe message
+     * that will be shown to the user.
+     *
+     * @param string     $safeMessage
+     * @param array      $messageData
+     * @param int        $code
+     * @param \Exception $previous
+     *
+     * @return CustomAuthenticationException
+     */
+    static public function createWithSafeMessage($safeMessage = '', array $messageData = array(), $code = 0, \Exception $previous = null)
+    {
+        $exception = new static($safeMessage, $code, $previous);
+        $exception->setSafeMessage($safeMessage, $messageData);
+
+        return $exception;
+    }
+
+    /**
+     * Set a message that will be shown to the user.
+     *
+     * @param string $messageKey  The message or message key
+     * @param array  $messageData Data to be passed into the translator
+     */
+    public function setSafeMessage($messageKey, array $messageData)
+    {
+        $this->messageKey = $messageKey;
+        $this->messageData = $messageData;
+    }
+
+    public function getMessageKey()
+    {
+        return null !== $this->messageKey ? $this->messageKey : parent::getMessageKey();
+    }
+
+    public function getMessageData()
+    {
+        return null !== $this->messageData ? $this->messageData : parent::getMessageData();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function serialize()
+    {
+        return serialize(array(
+            $this->messageKey,
+            $this->messageData,
+            parent::serialize(),
```

Comment:
I suggest putting the parent data first. This way, it is easier to add more fields in the future while preserving backward compatibility in the future (to keep the BC layer sane, new fields should be added at the end of array, which would then mean that the parent data goes in the middle of the object data)